include/fmt/format.h

```diff
@@ -4000,7 +4000,9 @@ struct formatter<T, Char, enable_if_t<detail::has_format_as<T>::value>>
   template <typename FormatContext>
   auto format(const T& value, FormatContext& ctx) const -> decltype(ctx.out()) {
     using base = formatter<detail::format_as_t<T>, Char>;
-    return base::format(format_as(value), ctx);
+    // format functions expect lvalue refs, not rvalues
+    auto val = format_as(value);
```

Comment:
Can it introduce an extra copy? Can `val` be a reference?

Reply:
If `format_as` returns a local variable of type `T` by value then no extra copy will be made. This was my thought of the predominant use case.

However for a `format_as` function signature like: `const std::string& format_as(T t)`. Let's say you have a lookup table of `std::string`s for `T`, then using `auto val` always copies, while `auto&& val` deduces the type of `val` as a `const &` so no copy is made.

https://godbolt.org/z/fqMrcY3c3

Thus I think your suggestion of using `auto&&` is a good one here! Will make the change